### DISCUSSION TITLE
How does qdrant handle data conversion of torch tensors? For example, if I have a tensor of type bfloat16

### DISCUSSION THREAD

**Discussion by SvenPVoigt** [qdrant/qdrant#7641] [Q&A]
How does qdrant do data conversions? Let's say I have a datatype of float16 and saving to a collection initialized with float32, what is the algorithm used to do the casting? Especially, if I have a torch.Tensor... is it first converted to a python list of floats (so using python type casting?) or is it directly sent to rust for conversion of float16 to float32?

```
emb = torch.tensor([0.1,0.2], dtype=torch.float16)
client.upsert(
    collection_name="test",
    points=[
        models.PointStruct(
            id="000",
            vector=emb,
        ),
    ],
)
```

Additionally, lots of embedding models are using the bfloat16 format. If I pass a vector of this format to qdrant what is it converted to and how? Not a lot of programming languages have native support. Is it first case to float32?

bfloat16 has a high dynamic range like float32 but decreased precision vs float16. If normalizing the vector to range [0,1], would it be best to directly cast bfloat16 to float16 or float32? I am assuming dynamic range doesn't really play that big of a role in range [0,1]? What does qdrant do? And does anyone have error metrics/MAP scores for comparing the casting types of bfloat16 to either float16/float32?

**Comment by generall** [comment]
qdrant doesn't work with the client's data representation. All requests are serialized via json or protobuf before it arrives at qdrant api.
So it doesn't matter how pytorch represents floats

**Reply by SvenPVoigt** [reply]
Ok. So casting to float32 then converting to JSON would preserve all the significant digits of the bfloat16 representation then. And then casting to float16 would happen internally to qdrant after it parses the JSON as float32?

Also, do you have any insights into the precision needed for L2 normalized vectors with values in the range [-1,1]? Will quantization increase errors?

**Reply by generall** [reply]
> Also, do you have any insights into the precision needed for L2 normalized vectors with values in the range [-1,1]?


i guess that depends on the distribution.

**Reply by generall** [reply]
> Will quantization increase errors?

yes

**Comment by akkupratap323** [comment]
There are two separate conversions here: **client transport** and **Qdrant storage**.

For the remote server, do not rely on a raw `torch.Tensor` being preserved. Convert it explicitly at the boundary:

```python
vector = embedding.detach().cpu().to(torch.float32).tolist()
point = models.PointStruct(id="000", vector=vector)
```

That conversion does not recover precision already absent from a BF16 embedding; it preserves the BF16 values in a representation the client/API can serialize. The request then carries numeric values over JSON, or protobuf `float` values over gRPC. Qdrant's default collection datatype is Float32, so the stored original vector is Float32 unless the collection was created with another `datatype`.

If the model output is BF16, casting **BF16 -> Float32** before serialization is the safe default. Casting BF16 -> Float16 first introduces another rounding/range decision for no transport benefit. L2 normalization largely removes overflow concerns, but it does not make precision irrelevant: nearest-neighbor ordering can still change when many candidates have very close scores.

Qdrant now lets the collection's original-vector datatype be configured as Float32 (default) or Float16, among others. That is distinct from Qdrant quantization:

- `datatype=FLOAT16` stores the original vector in half precision.
- scalar/binary/product quantization creates a separate compressed search representation; the original can remain available for rescoring.

So I would ingest as Float32 first, establish an exact-search quality baseline, then compare Float16 and/or quantization on your own query set. Measure Recall@K or nDCG against the Float32 exact baseline and test with your real score distribution. There is no model-independent MAP penalty: dimensionality, distribution, distance metric, and how close the candidates are all affect it. If you enable quantization, retrieve a larger candidate set and rescore with the original vectors to recover recall.

The current datatype and quantization behavior is documented here:
- https://qdrant.tech/documentation/manage-data/vectors/#datatypes
- https://qdrant.tech/documentation/manage-data/quantization/
